### PR TITLE
Nnembedding

### PR DESCRIPTION
**Pull request by mova**
adding support for training with embedded samples for zττ
cleaning up  `ml/write_dataset_config.py`
`ml/sum_training_weights.py`can now write class weights
`utils/multirun.sh`allows parallel execution if `PARALLEL=`is set
Minor bug fixes in  plot_nominal.py 

- `ml/write_dataset_config.py`
    - additional argument : `--training-z-estimation-method` for switching between the Z estimation method between `emb` (embedding) and `mc` (MonteCarlo)
    - the hardcoded options for each channel is removed and replaced by a logic
        - `estimationMethodAndClassMapGenerator()` simultaniusly builds up the maaping of the classes and a list of estimation methods (from the given era), depending on `channel`, `training-z-estimation-method`
        - channel objects from `shape_producer.channel` are loaded into a dict and selected according to channel and era
        - logic for selecting different QCD sideband cuts for "tt"
- `ml/create_training_dataset.sh`
    - passes new option `--training-z-estimation-method` to `ml/write_dataset_config.py`, emb is default

- `ml/sum_training_weights.py`
    - channels read from dataset_config_file
    - new argument `--write-weights` (bool): if true, writes new inverse class weight to `ml/$era_$channel_training.yaml`
- `ml/sum_training_weights.sh`
    - now takes the `dataset_config_file` instead of the channels
    - now takes a channel as second argument. If none is given, the script is iterated over all channels.
    - default setting: orders `sum_training_weights.py` do overwrite the class weights with the ones generated from `combined_training_dataset.root`
    - uses `utils/multirun.sh` to run 
- `utils/multirun.sh`
    - takes two arguments from the list and passes them to `run_procedure`
    - if PARALLEL is set two 1, the a given function is executed with a given argument list
    - otherwise the programms are executed sequentially
    - usage:
```
#method that takes ERA and channel 
function run_procedure () {
...
}

source utils/multirun.sh
genArgsAndRun run_procedure $ERA $CHANNEL 
```

### CODE REVIEW

File: ml/create_training_dataset.sh [hunk 1]
```diff
@@ -7,46 +7,67 @@ source utils/setup_cvmfs_sft.sh
...
+
+    # Reweight STXS stage 1 signals so that each stage 1 signal is weighted equally but
+    # conserve the overall weight of the stage 0 signal
+    #python ml/reweight_stxs_stage1.py \
+    #    ml/${SELERA}_${SELCHANNEL} \
+    #    ml/${SELERA}_${SELCHANNEL}/fold0_training_dataset.root \
+    #    ml/${SELERA}_${SELCHANNEL}/fold1_training_dataset.root
+}
+
+CHANNELS=$(if [[ -z $CHANNEL ]]; then echo "tt" "mt" "et"; else echo $CHANNEL; fi)
```
Comment (ArturAkh) [inline ml/create_training_dataset.sh]: What about `em` channel?

File: ml/create_training_dataset.sh [hunk 2]
```diff
@@ -7,46 +7,67 @@ source utils/setup_cvmfs_sft.sh
...
+ERAS=$(if [[ -z $ERA ]]; then echo "2016" "2017"; else echo $ERA; fi)
```
Comment (ArturAkh) [inline ml/create_training_dataset.sh]: To be most general, please also add `"2018"`

File: ml/write_dataset_config.py [hunk 3]
```diff
@@ -71,636 +75,180 @@ def main(args):
...
+    channel_qcd = copy.deepcopy(channel)
+
+    if args.channel != "tt":
+        ## os= opposite sign
+        channel_qcd.cuts.get("os").invert()
+    # Same sign selection for data-driven QCD
+    else:
+        channel_qcd.cuts.remove("tau_2_iso")
+        channel_qcd.cuts.add(
+            Cut("byTightIsolationMVArun2v1DBoldDMwLT_2<0.5", "tau_2_iso"))
```
Comment (ArturAkh) [inline ml/write_dataset_config.py]: Here, you would also need to distinguish between the years `2017` etc.

But the currently supported ID is (for all three years):

`byLooseIsolationMVArun2017v2DBoldDMwLT2017_2` (loose WP)
`byTightIsolationMVArun2017v2DBoldDMwLT2017_2` (tight WP)

Question to Sebastian W. and Stefan W.: shouldn't it be `byVLoose` WP?

File: ml/write_dataset_config.py [hunk 4]
```diff
@@ -71,636 +75,180 @@ def main(args):
...
+        if args.era=="2016":
+            channel_qcd.cuts.add(
+                Cut("byLooseIsolationMVArun2v1DBoldDMwLT_2>0.5", "tau_2_iso_loose"))
+        elif args.era=="2017":
```
Comment (ArturAkh) [inline ml/write_dataset_config.py]: For the latest implementation, you don't need to distinguish here between years (see commend above)

**Review by ArturAkh — CHANGES_REQUESTED** [review]
Hi Moritz,
Nice changes!
Please have a look at my small comments :)

Cheers,

Artur

File: ml/create_training_dataset.sh (hunk 1 above)
Comment (mova) [reply ml/create_training_dataset.sh]: done

File: ml/create_training_dataset.sh (hunk 2 above)
Comment (mova) [reply ml/create_training_dataset.sh]: done

**Review by ArturAkh — APPROVED** [review]
Hi Moritz,

Thanks for implementing my requests! :) From my side, your changes should be fine, let's wait for the comments of the others.

Cheers,

Artur

**Review by janekbechtel — APPROVED** [review]

**Pull request merged**